crates/viewer/re_viewer/tests/blueprint_test.rs

```diff
@@ -0,0 +1,142 @@
+use re_test_viewport::TestContextExt as _;
+use re_viewer_context::ViewClass as _;
+use re_viewport::ViewportUi;
+use re_viewport_blueprint::ViewBlueprint;
+
+fn log_test_data_and_register_views(test_context: &mut TestContext, scalars_count: usize) {
+    test_context.register_view_class::<re_view_dataframe::DataframeView>();
+    test_context.register_view_class::<re_view_bar_chart::BarChartView>();
+
+    let timeline_a = re_chunk::Timeline::new_sequence("timeline_a");
+    test_context.log_entity("scalar", |builder| {
+        builder.with_archetype(
+            RowId::new(),
+            [(timeline_a, 0)],
+            &re_types::archetypes::Scalars::single(scalars_count as f32),
+        )
+    });
+
+    let vector = (0..scalars_count).map(|i| i as f32).collect::<Vec<_>>();
+
+    test_context.log_entity("vector", |builder| {
+        builder.with_archetype(
+            RowId::new(),
+            TimePoint::STATIC,
+            &re_types::archetypes::BarChart::new(vector),
+        )
+    });
+}
+
+fn setup_viewport(test_context: &mut TestContext) {
+    let view_1 =
+        ViewBlueprint::new_with_root_wildcard(re_view_bar_chart::BarChartView::identifier());
+    let view_2 =
+        ViewBlueprint::new_with_root_wildcard(re_view_dataframe::DataframeView::identifier());
+
+    test_context.setup_viewport_blueprint(|ctx, blueprint| {
+        // Set the color override for the bar chart view.
+        let color_override = re_types::archetypes::BarChart::default().with_color([255, 144, 1]); // #FF9001
+        let override_path = re_viewport_blueprint::ViewContents::override_path_for_entity(
+            view_1.id,
+            &re_chunk::EntityPath::from("vector"),
+        );
+        ctx.save_blueprint_archetype(override_path.clone(), &color_override);
+
+        // Set the timeline for the dataframe view.
+        let query = re_view_dataframe::Query::from_blueprint(ctx, view_2.id);
+        query.save_timeline_name(ctx, &re_chunk::TimelineName::from("timeline_a"));
+
+        blueprint.add_views([view_1, view_2].into_iter(), None, None);
+    });
+}
+
+fn save_blueprint_to_file(test_context: &TestContext, path: &Path) {
+    test_context
+        .save_blueprint_to_file(path)
+        .expect("Failed to save blueprint to file.");
+}
+
+fn load_blueprint_from_file(test_context: &mut TestContext, path: &Path) {
+    let file = std::fs::File::open(path).expect("Failed to open blueprint file.");
+    let rbl_store =
+        re_entity_db::StoreBundle::from_rrd(file).expect("Failed to load blueprint store");
+    {
+        let mut lock = test_context.store_hub.lock();
+        let app_id = lock.active_app().expect("Missing active app").clone();
+        lock.load_blueprint_store(rbl_store, &app_id)
+            .expect("Failed to load blueprint store");
+    }
+
+    // Trigger recalculation of visualizable entities and blueprint overrides.
+    test_context.setup_viewport_blueprint(|_ctx, _blueprint| {});
```

Comment:
oof that's the annoying thing you've mentioned on yesterday standup where App is recalculating it all the time and `TestContext` is not? Feels like `TestContext` is in the wrong here, but I figure we can't do that where test context's run method lives or something of that shape?

Reply:
Yes, it is. :(
The new `App`-based harness work differently, it loads the current blueprint from the store at the start of each frame. The `TestContext` only does that when you explicitly call this function.